### PR TITLE
Feat/entproto printtemplate helper

### PR DESCRIPTION
**Pull request by rotemtam**
Notice the PR here is to feat/entproto-fieldmap, this is just for convenience once that is merged i will change it to master

### CODE REVIEW

File: entproto/cmd/protoc-gen-entgrpc/template.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+			}
+		}
+	}
+	if inToken {
+		return fmt.Errorf("entproto: corrupt template, must close parenthesis")
+	}
+	output = append(output, buf)
+	printer.P(output...)
+	return nil
+}
```
Comment (a8m) [inline entproto/cmd/protoc-gen-entgrpc/template.go:64]: Nice, you can achieve something similar with regexp:

```go
re := regexp.MustCompile(`(%\(.+?\))`)
out := re.ReplaceAllStringFunc(tmpl, func(s string) string {
	switch v := values[s]; v.(type) {
	case protogen.GoIdent:
		return g.QualifiedGoIdent(v) // The P(..) magic for Go identifiers. 
	default:
		return fmt.Sprint(v)
	}
}))
g.P(out)
```

File: entproto/cmd/protoc-gen-entgrpc/template.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+// Instead of:
+//	g.P("func ", svcName, "(", paramName, " string)")
+// We can use
+//	printTemplate(g, "func %(svcName)(%(paramName) string)
+func printTemplate(printer protogenPrinter, template string, values tmplValues) error {
+	var inToken bool
+	var buf string
+	var output []interface{}
+	for _, c := range template {
+		str := string(c)
```
Comment (a8m) [inline entproto/cmd/protoc-gen-entgrpc/template.go:33]: Why this conversion is needed? You can compare runes as well (e.g. `c == '('`)

File: entproto/cmd/protoc-gen-entgrpc/template.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+			}
+		} else {
+			if str == "%" {
+				inToken = true
+				output = append(output, buf)
+				buf = str
+			} else {
+				buf += str
+			}
+		}
```
Comment (a8m) [inline entproto/cmd/protoc-gen-entgrpc/template.go:56]:
```suggestion
		} else if str == "%" {
			inToken = true
			output = append(output, buf)
			buf = str
		} else {
			buf += str
		}
```

Or `switch` with `default`.

File: entproto/cmd/protoc-gen-entgrpc/template.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+// printTemplate is a utility function to make working with protogen have a more declarative interface.
+// It receives a protogenPrinter (in practice a *protogen.GeneratedFile), a template string with placeholder
+// formatted like "%(variableName)" and a tmplValues map containing the values that should be replaced when
+// the template is rendered.
+//
+// Instead of:
+//	g.P("func ", svcName, "(", paramName, " string)")
+// We can use
+//	printTemplate(g, "func %(svcName)(%(paramName) string)
+func printTemplate(printer protogenPrinter, template string, values tmplValues) error {
```
Comment (a8m) [inline entproto/cmd/protoc-gen-entgrpc/template.go:28]: This interface is not really needed. Instead, pass a print function instead. 

```go
func printTemplate(print func(...interface{}), ....) error

// Or:
type PrintFunc func(...interface{})

func printTemplate(print PrintFunc, ....) error 
```

File: entproto/cmd/protoc-gen-entgrpc/template.go (hunk 1 above)
Comment (rotemtam) [reply entproto/cmd/protoc-gen-entgrpc/template.go:56]: this is a really great suggestion. i ended up doing just that. my prime thought is that its much easier to change if we need to add more subtleties to the "template language"

Comment (a8m) [reply entproto/cmd/protoc-gen-entgrpc/template.go:56]: Great 😃

**Review by a8m — APPROVED** [review]

File: entproto/cmd/protoc-gen-entgrpc/template.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+//
+// Instead of:
+//	g.P("func New", svcName, "(p string) *", protogen.GoImportPath("context").Ident("Context"))
+// We can use
+//	printTemplate(g, "func New%(svcName)(p string) *%(ctx)", tmplValues{
+//		"svcName": "UserService",
+//		"ctx": protogen.GoImportPath("context").Ident("Context"),
+//	})
+func printTemplate(g *protogen.GeneratedFile, tmpl string, values tmplValues) error {
+	re := regexp.MustCompile(`(%\(.+?\))`)
```
Comment (a8m) [inline entproto/cmd/protoc-gen-entgrpc/template.go:37]: Just one comment - unlike Python, the `regexp` package doesn't cache these regular-expressions.
Move this variable to the top-level declarations:

```go
var re = regexp.MustCompile(`(%\(.+?\))`)

func printTemplate(g *protogen.GeneratedFile, tmpl string, values tmplValues) error {
	// ...
}
```

**Pull request merged**